gym/envs/registration.py

```diff
@@ -17,22 +17,49 @@
 
 from gym import error, logger
 
-# This format is true today, but it's *not* an official spec.
-# [username/](env-name)-v(version)    env-name is group 1, version is group 2
-#
-# 2016-10-31: We're experimentally expanding the environment ID format
-# to include an optional username.
-env_id_re: re.Pattern = re.compile(
-    r"^(?:(?P<namespace>[\w:-]+)\/)?(?P<name>[\w:.-]+)-v(?P<version>\d+)$"
-)
 
+def check_env_id(env_str: str) -> dict:
+    """Check environment ID string format.
 
-def env_id_from_parts(namespace: str, name: str, version: Union[int, str]) -> str:
+    This format is true today, but it's *not* an official spec.
+    [username/](env-name)-v(version)    env-name is group 1, version is group 2
+
+    2016-10-31: We're experimentally expanding the environment ID format
+    to include an optional username.
+    """
+    # First we try the strict pattern with the version number
+    pattern = re.compile(
+        r"^(?:(?P<namespace>[\w:-]+)\/)?(?P<name>[\w:.-]+)-v(?P<version>\d+)$"
+    )
+
+    match = pattern.fullmatch(env_str)
+    if match:
+        namespace, name, version = match.group("namespace", "name", "version")
+    else:
+        # Then we try a less strict pattern without the version number
+        pattern = re.compile(r"^(?:(?P<namespace>[\w:-]+)\/)?(?P<name>[\w:.-]+)$")
+        match = pattern.fullmatch(env_str)
+        if not match:
+            raise error.Error(
+                f"Malformed environment ID: {env_str}."
+                f"(Currently all IDs must be of the form {pattern}.)"
+            )
+        namespace, name = match.group("namespace", "name")
+        version = None
+    res = {"matched": match, "pattern": pattern, "id_parts": (namespace, name, version)}
```

Comment:
Is there a particular reason you were returning `pattern` (it doesn't seem to be used) and `id_parts`? It seems we could just return `match` and call `group` where needed. We're guaranteed to get a valid match out of this function so it's valid to just call `match.group('...')` for whatever we need.

Reply:
Could you also update the return type to be `re.Match`.

Reply:
For the `pattern` there was a reason before as it was used in error messages elsewhere, but then I moved the error messages inside the function, so indeed there's no reason anymore, good catch! Agree with the rest.